tools/ci-cdk/canary-lambda/src/latest/s3_canary.rs

```diff
@@ -82,38 +87,90 @@ pub async fn s3_canary(client: s3::Client, s3_bucket_name: String) -> anyhow::Re
         return Err(CanaryError(format!("presigned URL returned bad data: {:?}", response)).into());
     }
 
-    let mut result = Ok(());
-    match output.metadata() {
-        Some(map) => {
-            // Option::as_deref doesn't work here since the deref of &String is String
-            let value = map.get("something").map(|s| s.as_str()).unwrap_or("");
-            if value != METADATA_TEST_VALUE {
-                result = Err(CanaryError(format!(
+    let metadata_value = output
+        .metadata()
+        .and_then(|m| m.get("something"))
+        .map(String::as_str);
+    let result: anyhow::Result<()> = match metadata_value {
+        Some(value) => {
+            if value == METADATA_TEST_VALUE {
+                let payload = output
+                    .body
+                    .collect()
+                    .await
+                    .context("download s3::GetObject[2] body")?
+                    .into_bytes();
+                if std::str::from_utf8(payload.as_ref()).context("s3 payload")? == "test" {
+                    Ok(())
+                } else {
+                    Err(CanaryError("S3 object body didn't match what was put there".into()).into())
+                }
+            } else {
+                Err(CanaryError(format!(
                     "S3 metadata was incorrect. Expected `{}` but got `{}`.",
                     METADATA_TEST_VALUE, value
                 ))
-                .into());
+                .into())
             }
         }
-        None => {
-            result = Err(CanaryError("S3 metadata was missing".into()).into());
-        }
-    }
+        None => Err(CanaryError("S3 metadata was missing".into()).into()),
+    };
 
-    let payload = output
-        .body
-        .collect()
+    // Delete the test object
+    client
+        .delete_object()
+        .bucket(&s3_bucket_name)
+        .key(&test_key)
+        .send()
         .await
-        .context("download s3::GetObject[2] body")?
-        .into_bytes();
-    if std::str::from_utf8(payload.as_ref()).context("s3 payload")? != "test" {
-        result = Err(CanaryError("S3 object body didn't match what was put there".into()).into());
-    }
+        .context("s3::DeleteObject")?;
+
+    // Return early if the result is an error
+    result?;
+
+    // Put the test object
+    client
+        .put_object()
+        .bucket(&s3_mrap_bucket_arn)
+        .key(&test_key)
+        .body(ByteStream::from_static(b"test"))
+        .metadata("something", METADATA_TEST_VALUE)
+        .send()
+        .await
+        .context("s3::PutObject[MRAP]")?;
+
+    // Get the test object and verify it looks correct
+    let output = client
+        .get_object()
+        .bucket(&s3_mrap_bucket_arn)
+        .key(&test_key)
+        .send()
+        .await
+        .context("s3::GetObject[MRAP]")?;
```

Comment:
should we override the region or something to make sure we're really testing multi-region?

Reply:
Good idea. I've added that.